Added Rector and Drupal Check and fixed Drupal 10 compatibility.

AlexSkrypnyk opened the pull request:
- Added Rector
- Added Drupal Check
- Fixed code that was not compatible with Drupal 10
- Fixed compatibility with `Symfony\Process` in Drush Driver (copied from https://github.com/jhedstrom/DrupalDriver/pull/251 - sorry, had to keep it in this branch for downstream projects)

Please note that Rector and Drupal check are only scanning Drupal 8 driver.


<img width="990" alt="Build__767_-_jhedstrom_DrupalDriver_-_Travis_CI" src="https://user-images.githubusercontent.com/378794/194359882-f97ad32f-4b43-4a25-b5c3-a09dda495233.png">

pfrenssen commented on .travis.yml:
```diff
@@ -1,8 +1,6 @@
 
 php:
-  - 7.4
-  - 8.0
```
Any reason to stop testing on PHP 8.0? It is still supported until October 2023.

I think for as long as we declare in composer.json to be compatible with PHP 7.4 we should also keep testing on 7.4. Probably we shouldn't drop PHP 7.4 support unless we plan to release a new major version.

I suppose the `./vendor/bin/drupal-check` test only works on PHP 8.1, but we can still run the remainder of the suite on older PHP versions.

AlexSkrypnyk replied:
Drupal 10's requirement is to use 8.1. The CI may fail for 8.0 on D10

AlexSkrypnyk replied:
PHP8 did not work

```
Your requirements could not be resolved to an installable set of packages.
  Problem 1
    - drupal/core-recommended[10.0.0-beta1, ..., 10.0.0-beta2] require symfony/process ~v6.1.3 -> satisfiable by symfony/process[v6.1.3].
    - symfony/process[v6.1.0-BETA1, ..., v6.1.3] require php >=8.1 -> your php version (8.0.23) does not satisfy that requirement.
    - Root composer.json requires drupal/core-recommended ^10@beta -> satisfiable by drupal/core-recommended[10.0.0-beta1, 10.0.0-beta2].
The command "composer install" failed and exited with 2 during .
```

https://app.travis-ci.com/github/jhedstrom/DrupalDriver/jobs/585619255

I've reverted my change so that we are only testing on PHP 8.1

If there is a different way to test with PHP 8.0 - please let me know and I will update this PR.

pfrenssen commented on composer.json:
```diff
@@ -15,25 +15,34 @@
   },
   "scripts": {
     "test": [
       "composer validate --no-interaction",
       "parallel-lint src spec tests",
       "phpunit",
       "phpspec run -f pretty --no-interaction",
-      "phpcs --standard=./phpcs-ruleset.xml ."
-
+      "phpcs --standard=./phpcs-ruleset.xml --ignore=drupal .",
```
Let's not add the `--ignore` option inline. Let's instead update the ignored paths in `phpcs-ruleset.xml`:

```
<exclude-pattern>./drupal/*</exclude-pattern>
```

pfrenssen commented on src/Drupal/Driver/Cores/Drupal8.php:
```diff
@@ -385,7 +392,7 @@ public function getExtensionPathList() {
    * @param array $base_fields
    *   Base fields to be expanded in addition to user defined fields.
    */
-  public function expandEntityBaseFields($entity_type, \stdClass $entity, array $base_fields) {
+  public function expandEntityBaseFields($entity_type, $entity, array $base_fields) {
```
Why remove the `\stdClass` type hint? This is here to clarify that this is not a real entity but rather a "mocked" value object. It is literally a `\stdClass` object so the type hint is correct. Am I missing something?

AlexSkrypnyk replied:
This, as well as other changes, is a result of Rector's and drupal-check report. I don't remember what static analysis rule this was violating specifically, but it would not pass those tools.

We can revert and add an ignore comment for those tools to ignore this. 

Please advise

pfrenssen commented on src/Drupal/Driver/Fields/Drupal8/ImageHandler.php:
```diff
@@ -17,9 +17,7 @@ public function expand($values) {
-    $file = file_save_data(
-      $data,
-      'public://' . uniqid() . '.jpg');
+    $file = \Drupal::service('file.repository')->writeData($data, 'public://' . uniqid() . '.jpg');
```
This has been introduced in Drupal 9.3, so this breaks compatibility with Drupal 8 and Drupal 9 up to version 9.2.x

I am not opposed to this, but if this gets in we have to clearly stop supporting Drupal 9.2 and below. We need to add this to `composer.json`:

```
"conflict": {
    "drupal/core": ">=8.0 <9.3"
},
```

This will ensure that any projects stuck on an old version of Drupal for any reason will not accidentally break if they update DrupalDriver. They will remain on the last supported version.

AlexSkrypnyk replied:
added

pfrenssen commented on src/Drupal/Driver/Cores/Drupal8.php:
```diff
@@ -4,6 +4,7 @@
 
 use Drupal\Core\DrupalKernel;
 use Drupal\Core\Field\BaseFieldDefinition;
+use Drupal\Core\Routing\RouteObjectInterface;
```
This has been introduced in Drupal 9.1, so adding this will break compatibility with Drupal 8.x and 9.0. I think it's OK to introduce this but we need to indicate this incompatibility in our `composer.json` file, using the `conflict` directive I mentioned in another comment.

claudiu-cristea replied:
@AlexSkrypnyk @pfrenssen I've fixed this in https://github.com/jhedstrom/DrupalDriver/pull/253/files by hardcoding the constant value, so we keep compatibility with Drupal 8/9

AlexSkrypnyk commented:
@pfrenssen 
Thank you for reviewing this. Agree with all comments. I will address them shortly. Thanks

AlexSkrypnyk commented:
@pfrenssen 
I have addressed all the comments. Could you please review again. Thank you

jhedstrom commented:
re: testing on previous PHP versions

I think it is these lines that are causing those to not work:

```
    "drupal/core-composer-scaffold": "^10@beta",
    "drupal/core-recommended": "^10@beta",
```

They can probably either be removed, or switched to allow the older versions of Drupal as in the `require` section:

```
    "drupal/core-utility": "^8.4 || ^9 || ^10.@beta"
```

AlexSkrypnyk commented:
@jhedstrom 
could you please advise what do they have to be switched to?

jhedstrom commented:
I think they just need to allow 8.4 and 9 like we do for the `core-utility`:

```
    "drupal/core-composer-scaffold": "^8.4 || ^9 || ^10@beta",
    "drupal/core-recommended": "^8.4 || ^9 || ^10@beta",
```

(we could probably drop 8.x for now too, and then drop 9 when it is eol.)

AlexSkrypnyk commented:
thank you @jhedstrom 
I've updated the configs, but ran out of OSS credits on Travis CI (can we switch to Circle or GH actions?). Not sure how to test this without those credits. Probably have to wait till Nov 1.

jhedstrom commented:
@AlexSkrypnyk I've opened #254 to track that. It should be a relatively small effort.

Also, there's https://github.com/jhedstrom/DrupalDriver/issues/249 that might be relevant to the version changes happening here for core libraries.

AlexSkrypnyk commented:
@claudiu-cristea @jhedstrom 
I've updated this PR with the latest changes from `master` and updated CI to include D10. Could you please review. Thank you

jhedstrom commented:
Thanks!

Kingdutch commented:
@jhedstrom This PR introduced a conflict with older Drupal versions which seems like something that shouldn't happen in a `patch` release (but should be a major version release instead). This prevents upgrading in some instances.

See https://github.com/goalgorilla/open_social_dev/pull/51 for the conflict message that this causes.
